src/main/java/com/urbanairship/datacube/idservices/HBaseIdService.java

```diff
@@ -4,221 +4,181 @@
         final long id = WithHTable.increment(pool, counterTable, counterKey, cf, QUALIFIER, 1L);
-        if(log.isDebugEnabled()) {
+
+        if (log.isDebugEnabled()) {
             log.debug("Allocated new id " + id);
         }
-        
+
         final long maxId = LongMath.pow(2L, numIdBytes * 8);
-        if(id > maxId) {
-            // This dimension has no more IDs available. We'll throw an exception, but first we'll
-            // remove the "ALLOCATING" record so future attempts can fail quickly.
-            // If this CAS fails, there's another concurrent thread trying to allocate an ID for
-            // the same input, but there's nothing we can do about it.
-            WithHTable.checkAndDelete(pool, lookupTable, lookupKey, cf, QUALIFIER, 
-                    allocRecord, new Delete(lookupKey));
+        if (id > maxId) {
+            // This dimension has no more IDs available.
             throw new RuntimeException("Exhausted IDs for dimension " + dimensionNum);
-        }
-        if(id < 0) {
+        } else if (id < 0) {
             throw new RuntimeException("Somehow ID was less than zero. Weird!");
         }
-        
-        /*
-         * PHASE 3
-         * 
-         * We have a unique value to be used as an id value. Now we persist the mapping from
-         * input -> id. 
-         * 
-         * This write is done as a checkAndPut to handle the rare possibility that
-         * another thread preempted our allocation record and allocated an ID. This could 
-         * occur if our thread suffered a long GC pause or other slowness. It's critical that
-         * we don't replace an existing valid mapping, since we must have complete consistency
-         * of IDs: once one thread uses an input->id mapping, every other thread must use it
-         * forever.
-         */
+
+        // Then try to CAS it into the lookup table. If this fails someone beat us to it, and a
+        // get should succeed.
         final Put put = new Put(lookupKey);
-        byte[] allocatedRecord = ArrayUtils.addAll(new byte[] {(byte)Status.ALLOCATED.ordinal()}, 
-                Util.longToBytes(id));
+        byte[] allocatedRecord = ArrayUtils.addAll(HBaseIdService.ALLOCATED_BYTES, Util.longToBytes(id));
         put.add(cf, QUALIFIER, allocatedRecord);
-        boolean swapSuccess = WithHTable.checkAndPut(pool, lookupTable, lookupKey, cf, QUALIFIER, 
-                allocRecord, put);
-        if(swapSuccess) {
+
+        boolean swapSuccess = WithHTable.checkAndPut(pool, lookupTable, lookupKey, cf, QUALIFIER, EMPTY, put);
+        if (swapSuccess) {
             return Util.leastSignificantBytes(id, numIdBytes);
         } else {
-            log.warn("Concurrent allocators!?!? ID " + id + " will never be used");
-            // Recurse, try again.
-            return getId(dimensionNum, input, numIdBytes);
+            final Optional<byte[]> currentValue = getId(lookupKey, numIdBytes);
```

Comment:
I've made it pretty clear that that was a danger, and one that I think is acceptable in part because the impact is fairly well bounded by the fact that the race window is tight and bounded to the flusher concurrency (since threads don't retry).

 As we've learned, making a dimension with high cardinality and frequent id creation mapped is a terrible idea, and this would only affect someone mapping a high cardinality set into a conservative number of bits. I believe that since id exhaustion is explicitly checked for and fails without corruption documenting that fact is acceptable as far as correctness guarantees.